src/features/ci-cd/hooks/service-installs.ts

```diff
@@ -95,78 +96,82 @@ hooks.addPureHook('POST', 'resin', 'device', {
 		await createAppServiceInstalls(
 			rootApi,
 			request.values.belongs_to__application,
-			[deviceId],
+			[request.values.device],
 		);
 	},
 });
 
-hooks.addPureHook('PATCH', 'resin', 'device', {
-	POSTRUN: async (args) => {
-		const affectedIds = args.request.affectedIds!;
-
-		// We need to delete all service_install resources for the current device and
-		// create new ones for the new application (if the device is moving application)
+hooks.addPureHook('PATCH', 'resin', 'device_application', {
+	POSTRUN: async ({ api, request }) => {
+		const affectedIds = request.affectedIds!;
 		if (
-			args.request.values.belongs_to__application != null &&
-			affectedIds.length !== 0
+			affectedIds.length === 0 ||
+			request.values.should_be_running__release === undefined
 		) {
-			await args.api.delete({
-				resource: 'service_install',
-				options: {
-					$filter: {
-						device: { $in: affectedIds },
+			return;
+		}
+
+		const deviceApps = (await api.get({
+			resource: 'device_application',
+			options: {
+				$select: 'device',
+				$filter: {
+					id: {
+						$in: affectedIds,
 					},
 				},
-			});
-			await createAppServiceInstalls(
-				args.api,
-				args.request.values.belongs_to__application,
-				affectedIds,
-			);
+			},
+		})) as Array<{ device: { __id: number } }>;
+
+		if (deviceApps.length === 0) {
+			return;
 		}
-	},
-});
 
-hooks.addPureHook('PATCH', 'resin', 'device', {
-	POSTRUN: async ({ api, request }) => {
-		const affectedIds = request.affectedIds!;
-		if (
-			request.values.should_be_running__release !== undefined &&
-			affectedIds.length !== 0
-		) {
-			// If the device was preloaded, and then pinned, service_installs do not exist
-			// for this device+release combination. We need to create these
-			if (request.values.should_be_running__release != null) {
-				await createReleaseServiceInstalls(api, affectedIds, {
-					id: request.values.should_be_running__release,
-				});
-			} else {
-				const devices = (await api.get({
-					resource: 'device',
-					options: {
-						$select: ['id', 'belongs_to__application'],
-						$filter: {
-							id: { $in: affectedIds },
+		const deviceIds = deviceApps.map(({ device }) => device.__id);
+
+		// If the device was preloaded, and then pinned, service_installs do not exist
+		// for this device+release combination. We need to create these
+		if (request.values.should_be_running__release != null) {
+			await createReleaseServiceInstalls(api, deviceIds, {
+				id: request.values.should_be_running__release,
+			});
+		} else {
+			const devices = (await api.get({
+				resource: 'device',
+				options: {
+					$select: 'id',
+					$expand: {
+						device_application: {
+							$select: ['belongs_to__application'],
```

Comment:
How about:
```suggestion
			const devices = (await api.get({
				resource: 'device_application',
				options: {
					$select: ['device', 'belongs_to__application'],
```
This removes the need for a JOIN and should also allow us to avoid doing the `[0]` in the `devicesByApp[groupId][0]` and also drop the `?` from `?.belongs_to__application.__id`?